Bottom: 9c310072b521ffea18593d0afeb11111a4396d44
Top:    081675de23cb221d7bab8087b3be8736bbaf4dd0
Author: Jason Wang <jasowang@redhat.com>
Date:   2016-05-10 21:21:30 +0800

Refresh of circular-buffer

---

diff --git a/drivers/net/tun.c b/drivers/net/tun.c
index acd85a8..475cbdb 100644
--- a/drivers/net/tun.c
+++ b/drivers/net/tun.c
@@ -1545,16 +1545,18 @@ static ssize_t tun_do_read(struct tun_struct *tun, struct tun_file *tfile,
 
 	if (tun->flags & IFF_TX_RING) {
 		unsigned head, tail;
+		struct tun_desc *desc;
 
 		spin_lock(&tfile->rlock);
-		/* Read index before reading contents at that index. */
-		head = smp_load_acquire(&tfile->head);
+		head = ACCESS_ONCE(tfile->head);
 		tail = tfile->tail;
 
 		if (CIRC_CNT(head, tail, TUN_RING_SIZE) >= 1) {
-			struct tun_desc *desc = &tfile->tx_descs[tail];
+			/* read tail before reading descriptor at tail */
+			smp_rmb();
+			desc = &tfile->tx_descs[tail];
 			skb = desc->skb;
-			/* Finish reading descriptor before incrementing tail. */
+			/* read descriptor before incrementing tail. */
 			smp_store_release(&tfile->tail,
 					  (tail + 1) & TUN_RING_MASK);
 		} else {
